server/datastore/mysql/hosts.go

```diff
@@ -1086,3 +1105,68 @@ func (d *Datastore) CleanupExpiredHosts(ctx context.Context) error {
 	}
 	return nil
 }
+
+func (d *Datastore) ListHostDeviceMapping(ctx context.Context, id uint) ([]*fleet.HostDeviceMapping, error) {
+	stmt := `
+    SELECT
+      id,
+      host_id,
+      email,
+      source
+    FROM
+      host_emails
+    WHERE
+      host_id = ?
+    ORDER BY
+      email, source`
+
+	var mappings []*fleet.HostDeviceMapping
+	err := sqlx.SelectContext(ctx, d.reader, &mappings, stmt, id)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "select host emails by host id")
+	}
+	return mappings, nil
+}
+
+func (d *Datastore) ReplaceHostDeviceMapping(ctx context.Context, hid uint, mappings []*fleet.HostDeviceMapping) error {
+	for _, m := range mappings {
+		if hid != m.HostID {
+			return ctxerr.Errorf(ctx, "host device mapping are not all for the provided host id %d, found %d", hid, m.HostID)
+		}
+	}
+
+	// the following SQL statements assume a small number of emails reported
+	// per host.
```

Comment:
Nit: Just curious, is this faster than making `INSERT ... ON DUPLICATE KEY UPDATE updated_at=CURRENT_TIMESTAMP` (with `(host_id, email, source)` defined as `UNIQUE`)?

Even if the entries are a few, these DELETEs+INSERTs will be called for all hosts.

As Tomas said, load testing will tell :)

Reply:
Ah, nvm we also need to delete them if there are no new mappings 👍